src/main/java/org/candlepin/model/PoolCurator.java

```diff
@@ -942,97 +902,49 @@ public List<Pool> getBySubscriptionIds(String ownerId, Collection<String> subIds
      *        is the Entitlement just created or modified.
      * @return Pools with too many entitlements for their new quantity.
      */
-    @SuppressWarnings("unchecked")
     public List<Pool> getOversubscribedBySubscriptionIds(String ownerId, Map<String, Entitlement> subIdMap) {
-        List<Criterion> subIdMapCriteria = new ArrayList<>();
-        Criterion[] exampleCriteria = new Criterion[0];
-        for (Entry<String, Entitlement> entry : subIdMap.entrySet()) {
-            SimpleExpression subscriptionExpr = Restrictions.eq("sourceSub.subscriptionId", entry.getKey());
-            Criterion srcEntIsNull = Restrictions.isNull("sourceEntitlement");
-            Criterion srcEntEquals = Restrictions.eq("sourceEntitlement", entry.getValue());
+        CriteriaBuilder cb = getEntityManager().getCriteriaBuilder();
+        CriteriaQuery<Pool> cq = cb.createQuery(Pool.class);
+        Root<Pool> pool = cq.from(Pool.class);
+        Join<Pool, Subscription> sourceSub = pool.join(Pool_.SOURCE_SUBSCRIPTION);
 
-            Junction subscriptionJunction = Restrictions.and(subscriptionExpr)
-                .add(Restrictions.or(srcEntIsNull, srcEntEquals));
-
-            subIdMapCriteria.add(subscriptionJunction);
+        List<Predicate> predicates = new ArrayList<>();
+        predicates.add(cb.equal(pool.get(Pool_.OWNER).get(Owner_.ID), ownerId));
+        predicates.add(cb.ge(pool.get(Pool_.QUANTITY), 0L));
+        predicates.add(cb.gt(pool.get(Pool_.CONSUMED), pool.get(Pool_.QUANTITY)));
+
+        List<Predicate> orPredicates = new ArrayList<>();
+        for (Map.Entry<String, Entitlement> entry : subIdMap.entrySet()) {
```

Comment:
With a large enough map, we could run into the query parameter limit. We should gate against that in some capacity, either with a check at the beginning or partitioning around the map.

Reply:
done